implementations/rust/ockam/ockam_core/src/routing/address.rs

```diff
@@ -92,38 +131,52 @@ impl Address {
     ///
     /// See type documentation for more detail
     pub fn from_string<S: Into<String>>(s: S) -> Self {
+        match s.into().parse::<Address>() {
+            Ok(a) => a,
+            Err(e) => {
+                panic!("Invalid address string {}", e)
+            }
+        }
+    }
+    /// Generate a random address with a specific type
+    pub fn random(tt: u8) -> Self {
+        Self { tt, ..random() }
+    }
+}
+impl core::str::FromStr for Address {
+    type Err = AddressParseError;
+    /// Parse an address from a string
+    ///
+    /// See type documentation for more detail
+    fn from_str(s: &str) -> Result<Address, Self::Err> {
         let buf: String = s.into();
         let mut vec: Vec<_> = buf.split('#').collect();
 
         // If after the split we only have one element, there was no
         // `#` separator, so the type needs to be implicitly `= 0`
-        let (tt, inner) = if vec.len() == 1 {
-            (0, vec.remove(0).as_bytes().to_vec())
+        if vec.len() == 1 {
+            Ok(Address {
+                tt: 0,
+                inner: vec.remove(0).as_bytes().to_vec(),
+            })
         }
         // If after the split we have 2 elements, we extract the type
         // value from the string, and use the rest as the address
         else if vec.len() == 2 {
-            let tt = match str::parse(vec.remove(0)) {
-                Ok(tt) => tt,
-                Err(e) => {
-                    panic!("Failed to parse address type: '{}'", e);
-                }
-            };
-
-            (tt, vec.remove(0).as_bytes().to_vec())
+            match str::parse(vec.remove(0)) {
+                Ok(tt) => Ok(Address {
+                    tt,
+                    inner: vec.remove(0).as_bytes().to_vec(),
+                }),
+                Err(e) => Err(AddressParseError::new(AddressParseErrorKind::InvalidAddr(
```

Comment:
(part of my `s/InvalidAddr/InvalidType/` nit)
```suggestion
                Err(e) => Err(AddressParseError::new(AddressParseErrorKind::InvalidType(
```